core/src/player.rs

```diff
@@ -1184,12 +1189,19 @@ impl Player {
             let needs_render = if events.is_empty() {
                 false
             } else {
+                let mut refresh = false;
                 for (object, event) in events {
-                    if !object.as_displayobject().removed() {
+                    let display_object = object.as_displayobject();
+                    if !display_object.removed() {
                         object.handle_clip_event(context, event);
                     }
+                    if display_object.as_avm1_button().is_some()
+                        || display_object.as_avm2_button().is_some()
+                    {
+                        refresh = true;
+                    }
```

Comment:
1) Only button events should trigger redraws, so I think this can check if `event.is_button_event()` is true first of all. This handles the case of a movieclip with button event handlers, because they should act similarly even though they aren't a button. This also takes care of ignoring the `MouseMove` case.

2) If it's a movieclip, ideally we want to only trigger a re-render if the clip actually cares about button events (it's in "button mode"). So we can also check `MovieClip::is_button_mode`. Avm1Button and Avm2Button are basically always in button mode.

So something like this:
```
if event.is_button_event() {
    if display_object.as_avm1_button().is_some() || display_object.as_avm2_button().is_some() || display_object.as_movie_clip().map(|mc| mc.is_button_mode()).unwrap_or_default() {
    refresh = true
   }
}
```
Possibly we want to make `is_button_mode` a method on `DisplayObject`, or make have `DisplayObject::handle_clip_event` tweak `context.needs_render` itself. But that can be a future PR to keep this simple.

(disclaimer: haven't tested any of the above :-) )

Reply:
Sample SWF:
[mc-button-event.zip](https://github.com/ruffle-rs/ruffle/files/8750373/mc-button-event.zip)